File: haskell-rest/CoinApi/Types/Orderbook.hs
{-# LANGUAGE OverloadedStrings #-}

module CoinApi.Types.Orderbook where

import CoinApi.Types.Internal

data Pair = Pair { price :: !Scientific
                 , size  :: !Int }
  deriving (Show, Eq)

instance FromJSON Pair where
  parseJSON = withObject "Bid/ask" $ \o -> Pair <$> o .: "price" <*> o .: "size"

data Orderbook = Orderbook { symbol_id :: !Text
                           , time_exchange :: !UTCTime
                           , time_coinapi :: !UTCTime
                           , asks :: [Pair]
                           , bids :: [Pair] }
  deriving (Show, Eq)

instance FromJSON Orderbook where
  parseJSON = withObject "Orderbook" $ \o -> Orderbook
                                             <$> o .: "symbol_id"
                                             <*> fmap fromTime (o .: "time_exchange")
                                             <*> fmap fromTime (o .: "time_coinapi")
                                             <*> o .: "asks"
                                             <*> o .: "bids"

File: csharp-ws/CoinApi.WEBSOCKET/CoinApi.WEBSOCKET.V1/DataModels/OrderBook.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CoinApi.WEBSOCKET.V1.DataModels
{
    public struct OrderBook
    {
        public bool? is_snapshot { get; set; }
        public DateTime time_exchange { get; set; }
        public DateTime time_coinapi { get; set; }
        public object[] asks { get; set; }
        public object[] bids { get; set; }
    }
}
